Price Indexer Performance Issue With Out of Stock Products

brucemead opened the pull request:
### Description (*)
Within the reindexing process there is a bug which results in a delete query to be called for every out of stock product during a full or partial reindex.

### Steps to reproduce (*)
1. Ensure that display out of stock products option is disabled
2. `php bin/magento indexer:reindex catalog_product_price`
3. `php bin/magento cron:run --group=index`

### Fixed Issues (if relevant)
1. magento/magento2#24414: Price Indexer Performance Issue With Out of Stock Products

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

brucemead committed:
Price Indexer Performance Issue With Out of Stock Products
- Relates to issue https://github.com/magento/magento2/issues/24414

aleron75 commented on app/code/Magento/CatalogInventory/Model/Indexer/ProductPriceIndexFilter.php:
```diff
@@ -104,6 +104,14 @@ public function modifyPrice(IndexTableStructure $priceTable, array $entityIds =
             $select->where('stock_item.use_config_manage_stock = 0 AND stock_item.manage_stock = 1');
         }
 
+        if ($entityIds !== null) {
```
I think the null check is not needed and can be removed, simplifying code: the `$entityIds` parameter has declared type so in case the `$entityIds` parameter of `modifyPrice()` function is `null`, PHP would throw a fatal error.

sidolov reviewed (changes requested):
@brucemead please, apply changes suggested by @aleron75

sidolov commented on app/code/Magento/CatalogInventory/Model/Indexer/ProductPriceIndexFilter.php:
```diff
@@ -104,6 +104,14 @@ public function modifyPrice(IndexTableStructure $priceTable, array $entityIds =
         }
 
+        if ($entityIds !== null) {
+            if (count($entityIds) > 1) {
```
Why we should count elements here? $entityIds always array, let's use `$select->where('stock_item.product_id in (?)', $entityIds)` in all cases if array is not empty

brucemead replied:
Yeah fair point I copied the implementation from here `Magento\Catalog\Model\ResourceModel\Product\Indexer\Price\Query\BaseFinalPrice::getQuery`

southerncomputer commented:
might be faster if you intval $entityids too!

also pull against the same thing in vendor/magento/module-catalog/Model/ResourceModel/Product/Indexer/Price/Query/BaseFinalPrice.php as it uses this same routine but with a between min/max(entityIds) which is horribly wrong to do for a list of 100 entityids (think 2000 and 2Million) over and over again!

sidolov commented:
@brucemead , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

brucemead committed:
Price Indexer Performance Issue With Out of Stock Products
- Relates to issue https://github.com/magento/magento2/issues/24414
- Made changes based on feedback on PR

brucemead commented:
@sidolov I've pushed the suggested changes now. Sorry for the slow turn around I've been on leave recently.

Thanks,
Bruce

aleron75 commented on app/code/Magento/CatalogInventory/Model/Indexer/ProductPriceIndexFilter.php:
```diff
@@ -104,6 +104,8 @@ public function modifyPrice(IndexTableStructure $priceTable, array $entityIds =
             $select->where('stock_item.use_config_manage_stock = 0 AND stock_item.manage_stock = 1');
         }
 
+        $select->where('stock_item.product_id in (?)', $entityIds);
```
I may be wrong but I'm not sure this is what @sidolov intended.

He probably meant that it's not necessary to distinguish between empty or not empty array when building the `$select->where('...IN...', ...)` because if the array is empty this will produce a `WHERE stock_item.product_id in (NULL)` statement that won't return any value.

And that's the point, if you always apply this statement without checking whether the array is empty, an empty array will always produce the above statement and won't return any result, that's not the expected behavior.

I think the code should be:
```
if (count($entityIds)) {
    $select->where('stock_item.product_id in (?)', $entityIds);
}
```

sidolov committed:
Added check for empty array

magento-engcom-team commented:
Hi @sidolov, thank you for the review.
[ENGCOM-5910](https://jira.corp.magento.com/browse/ENGCOM-5910) has been created to process this Pull Request
:eight_spoked_asterisk: @sidolov, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

magento-engcom-team commented:
Hi @aleron75, thank you for the review.
[ENGCOM-5910](https://jira.corp.magento.com/browse/ENGCOM-5910) has been created to process this Pull Request

brucemead committed:
Price Indexer Performance Issue With Out of Stock Products
- Relates to issue https://github.com/magento/magento2/issues/24414
- Made changes based on feedback on PR
- Resolving whitespace issue https://testing-service.magento-community.engineering/reports/magento/magento2/pull/24415/56e6453d1da3a1546212c6c35f78d77d/Statics/allure-report-ce/index.html#categories/479362f77c9377ec388bccba15830ad3/997602107dd59e40/

brucemead committed:
Merge remote-tracking branch 'origin/fix/catalog-inventory-price-indexer-performance-fix' into fix/catalog-inventory-price-indexer-performance-fix

# Conflicts:
#	app/code/Magento/CatalogInventory/Model/Indexer/ProductPriceIndexFilter.php

brucemead committed:
Price Indexer Performance Issue With Out of Stock Products
- Relates to issue https://github.com/magento/magento2/issues/24414
- Resolving whitespace issue https://testing-service.magento-community.engineering/reports/magento/magento2/pull/24415/56e6453d1da3a1546212c6c35f78d77d/Statics/allure-report-ce/index.html#categories/479362f77c9377ec388bccba15830ad3/997602107dd59e40/

brucemead commented:
@sidolov Just pushed a fix for the white space issue that was flagged here https://testing-service.magento-community.engineering/reports/magento/magento2/pull/24415/56e6453d1da3a1546212c6c35f78d77d/Statics/allure-report-ce/index.html#categories/479362f77c9377ec388bccba15830ad3/997602107dd59e40/retries

engcom-Alfa commented:
:heavy_check_mark: QA Passed

engcom-Alfa commented:
:heavy_check_mark: QA Passed

Nazar65 committed:
Cover changes with unit test.
